src/server/sample_data/routes/install.js

```diff
@@ -49,87 +49,95 @@ export const createInstallRoute = () => ({
+              settings: {
+                index: {
+                  number_of_shards: 1,
+                  number_of_replicas: 0
+                }
+              },
+              mappings: {
+                _doc: {
+                  properties: dataIndexConfig.fields
+                }
+              }
+            }
+          };
+          await callWithRequest(request, 'indices.create', createIndexParams);
+        } catch (err) {
+          const errMsg = `Unable to create sample data index "${index}", error: ${err.message}`;
+          server.log(['warning'], errMsg);
+          return reply(errMsg).code(err.status);
         }
 
-        const createResults = await request.getSavedObjectsClient().bulkCreate(sampleDataset.savedObjects, { overwrite: true });
-        const errors = createResults.saved_objects.filter(savedObjectCreateResult => {
-          return savedObjectCreateResult.hasOwnProperty('error');
-        });
-        if (errors.length > 0) {
-          server.log(['warning'], `sample_data install errors while loading saved objects. Errors: ${errors.join(',')}`);
-          return reply(`Unable to load kibana saved objects, see kibana logs for details`).code(403);
+        const bulkInsert = async (docs) => {
+          const bulk = [];
+          docs.forEach(doc => {
+            bulk.push(insertCmd);
+            bulk.push(updateTimestamps(doc, dataIndexConfig));
+          });
+          const resp = await callWithRequest(request, 'bulk', { body: bulk });
+          if (resp.errors) {
+            server.log(
+              ['warning'],
+              `sample_data install errors while bulk inserting. Elasticsearch response: ${JSON.stringify(resp, null, '')}`);
+            return Promise.reject(new Error(`Unable to load sample data into index "${index}", see kibana logs for details`));
+          }
+        };
+
+        try {
+          const count = await loadData(dataIndexConfig.dataPath, bulkInsert);
+          counts[index] = count;
+        } catch (err) {
+          server.log(['warning'], `sample_data install errors while loading data. Error: ${err}`);
+          return reply(err.message).code(500);
         }
+      }
 
-        return reply({ docsLoaded: count, kibanaSavedObjectsLoaded: sampleDataset.savedObjects.length });
+      const createResults = await request.getSavedObjectsClient().bulkCreate(sampleDataset.savedObjects, { overwrite: true });
+      const errors = createResults.saved_objects.filter(savedObjectCreateResult => {
+        return savedObjectCreateResult.hasOwnProperty('error');
       });
+      if (errors.length > 0) {
+        server.log(['warning'], `sample_data install errors while loading saved objects. Errors: ${errors.join(',')}`);
+        return reply(`Unable to load kibana saved objects, see kibana logs for details`).code(403);
+      }
+
+      return reply({ elasticsearchIndicesCreated: counts, kibanaSavedObjectsLoaded: sampleDataset.savedObjects.length });
```

Comment:
The handler is ~100 lines long. I don't always mind long functions / methods, but in this case, it was a bit hard to follow.

It seems like the handler is doing all of the following:

- updateTimestamps
- deleteOldSampleDataIndices
- createNewSampleDataIndices
- hydrateSampleDataIndices
- upsertSavedObjects
- Various forms of error handling / reporting

This would probably be improved by breaking it apart into smaller functions that are easier to digest at a glance.

Reply:
I moved all of the insert logic into a function `insertDataIntoIndex`. This greatly reduces the handler function size. I don't want to create too many small functions because most of the interesting interactions are around error handling and properly replying to the request.

Reply:
Hm. I'm not sure that the `insertDataIntoIndex` alone helps much. I'll leave it up to you whether you think it's worth refactoring further, going with the current factoring of `insertDataIntoIndex`, or just going with the original PR. It was mostly a code-clarity comment, so I'll leave it to your disgression.

Reply:
There is no way to loop through all of the data at once because it is never in memory at a single time. The data is streamed and processed in chunks. `loadData` streams the data from the compressed file. When the number of records reaches a threshold - then they are inserted into elasticsearch with the bulk API. Before insertion, the chunk of documents are processed. This processing includes translating timestamps to the current time.